CANTroller2/src/qpid.h

```diff
@@ -447,12 +436,153 @@ float QPID::GetOutputSum() { return outputSum; }
 float QPID::GetOutputMin() { return outMin; }  // Soren
 float QPID::GetOutputMax() { return outMax; }  // Soren
 float QPID::GetOutputRange() { return outMax - outMin; }  // Soren
+float QPID::GetCenter() { return center; }  // Soren
 uint8_t QPID::GetMode() { return static_cast<uint8_t>(mode); }
 uint8_t QPID::GetDirection() { return static_cast<uint8_t>(action); }
 uint8_t QPID::GetPmode() { return static_cast<uint8_t>(pmode); }
 uint8_t QPID::GetDmode() { return static_cast<uint8_t>(dmode); }
 uint8_t QPID::GetAwMode() { return static_cast<uint8_t>(iawmode); }
 uint8_t QPID::GetCentMode() { return static_cast<uint8_t>(centmode); }  // Soren
-float QPID::GetCenter() { return center; }  // Soren
 
+class IdleControl {  // Soren - To allow creative control of PID targets in case your engine idle problems need that.
+  public:
+    enum class idlemodes : uint32_t { direct, control, minimize, num_idlemodes };  // direct: disable idle management.  control: soft landing to idle rpm.  minimize: attempt to minimize idle to edge of instability
+  protected:
+    enum targetstates : uint32_t { driving, droptohigh, droptolow, idling, stallpoint };
+    // String modenames[3] = { "direct", "cntrol", "minimz" };
+    // String statenames[4] = { "drivng", "tohigh", "tolow", "tostal" };
+    targetstates runstate, laststate;
+    idlemodes idlemode;
+    float targetlast_rpm, idle_rpm, idlehigh_rpm, idlehot_rpm, idlecold_rpm, stallpoint_rpm, dynamic_rpm, temphot_f, tempcold_f;
+    float margin_rpm = 10; float idle_absmax_rpm = 1000.0;  // High limit of idle speed adjustability
+    float* target_rpm; float* measraw_rpm; float* measfilt_rpm; float* coolant_f;
+    bool we_just_changed_states = true;
+    uint32_t settletime_us;
+    Timer settleTimer;
+  public:
+    IdleControl (float* target_rpm_arg, float* measraw_rpm_arg, float* measfilt_rpm_arg, float* coolant_f_arg,  // Variable references: idle target, rpm raw, rpm filt, Engine temp

```

Comment:
Sometimes when I have a lot of args, I split them out into a struct. This way I can mess around with the args order and count and not have to retype all of the uses of the thing. This is just a personal choice thing, what you have already works great
eg
```
struct IdleControlConfig {
    float* target_rpm;
    float* measraw_rpm;
    // ... other members
};

IdleControl(const IdleControlConfig& config);
```